workos/user_management.py

```diff
@@ -316,6 +319,72 @@ def delete_organization_membership(self, organization_membership_id):
             token=workos.api_key,
         )
 
+    def get_authorization_url(
+        self,
+        redirect_uri,
+        connection_id=None,
+        organization_id=None,
+        provider=None,
+        domain_hint=None,
+        login_hint=None,
+        state=None,
+    ):
+        """Generate an OAuth 2.0 authorization URL.
+
+        The URL generated will redirect a User to the Identity Provider configured through
+        WorkOS.
+
+        Kwargs:
+            redirect_uri (str) - A Redirect URI to return an authorized user to.
+            connection_id (str) - The connection_id connection selector is used to initiate SSO for a Connection.
+                The value of this parameter should be a WorkOS Connection ID. (Optional)
+            organization_id (str) - The organization_id connection selector is used to initiate SSO for an Organization.
+                The value of this parameter should be a WorkOS Organization ID. (Optional)
+            provider (str) - The provider connection selector is used to initiate SSO using an OAuth-compatible provider.
+                Currently, the supported values for provider are 'authkit', 'GoogleOAuth' and 'MicrosoftOAuth'. (Optional)
+            domain_hint (str) - Can be used to pre-fill the domain field when initiating authentication with Microsoft OAuth,
+                or with a GoogleSAML connection type. (Optional)
+            login_hint (str) - Can be used to pre-fill the username/email address field of the IdP sign-in page for the user,
+                if you know their username ahead of time. Currently, this parameter is supported for OAuth, OpenID Connect,
+                OktaSAML, and AzureSAML connection types. (Optional)
+            state (str) - An encoded string passed to WorkOS that'd be preserved through the authentication workflow, passed
+                back as a query parameter. (Optional)
+
+        Returns:
+            str: URL to redirect a User to to begin the OAuth workflow with WorkOS
+        """
+        params = {
+            "client_id": workos.client_id,
+            "redirect_uri": redirect_uri,
+            "response_type": RESPONSE_TYPE_CODE,
+        }
+
+        if connection_id is None and organization_id is None and provider is None:
+            raise ValueError(
+                "Incomplete arguments. Need to specify either a 'connection_id', 'organization_id', or 'provider_id'"
+            )
+
+        if connection_id is not None:
+            params["connection_id"] = connection_id
+        if organization_id is not None:
+            params["organization_id"] = organization_id
+        if provider is not None:
+            params["provider"] = provider
```

Comment:
We can do the same is-valid-provider check that this does: https://github.com/workos/workos-python/blob/91a24b315767ba86a19b373aca17ec1fc5608714/workos/sso.py#L87-L101

Since it's a parameter that's used as input to the API, any new additions would not result in SDK-breaking changes; existing SDKs will simply not be able to use new features, which is fine. (It's when an enum value that comes out of the API as a response that becomes problematic if checked at runtime.)

Reply:
Just to confirm here - this is a check against the ConnectionType enum, defined here: https://github.com/workos/workos-python/blob/main/workos/utils/connection_types.py.

I believe unless the input is a defined enum value, we will throw an error: https://github.com/workos/workos-python/blob/d2e71041a42f152c1da8e9779499f75fa3e3b3e1/tests/test_sso.py#L207

The supported values for provider are `GoogleOAuth`, `MicrosoftOAuth`, and `authkit`. I can add `authkit` to the enum, but I don't think any new values would be supported unless added to the enum.

I actually question whether we should have this check for the SSO method, given that the ConnectionType spans more values than are actually supported in the `provider`, so it would be better to make a different enum or just go with a string.

For reference, I used the Node SDK as reference, and we just have it as a string type there: https://github.com/workos/workos-node/blob/abb683ccf55c79b7bb48639ed591fd4241f12f1e/src/user-management/interfaces/authorization-url-options.interface.ts#L7

Reply:
Ah I see. The API used to (and I guess still does) accept connection types for `provider` despite not publicly documenting that it does, which makes sense. I don't think we are cutting a new major version for the Python SDK so it doesn't make sense to change the SSO API.

It would make sense if we can have something like a `ProviderType` so that we can check against valid values for the User Management authorize API. And since `authkit` is not actually a Connection type, it wouldn't make sense to add it as a value there.

When we cut a major version for this SDK, we can change the SSO authorize url behavior as well so not all connection types can be passed into `provider`.

> For reference, I used the Node SDK as reference, and we just have it as a string type there.

It'd be nice if we had types there too. Arguably, it'd be a breaking SDK change if a provider value that was invalid suddenly became valid. I.e. if we add a new provider type in the API in the future, the SDK would not need a change, but previous failing requests would now succeed. Realistically, no one would care, but having enums for inputs into the API is always nice.

Reply:
Sounds good - the plan is to cut a major version, so I'll go ahead and create a ProviderType to use as an input for both SSO and UserManagement methods 👍